### PR TITLE
L0 parallel code generation and execution for Taxi queries.

### PR DESCRIPTION
**Pull request by kurapov-peter**
This PR enables simple group-by-agg queries multithreaded execution for the L0 path. It allows for executing all NYC Taxi queries to run on the GPU. No shared memory is used currently. Group initialization happens on the CPU.

### CODE REVIEW

File: omniscidb/L0Mgr/L0Mgr.cpp [hunk 1]
```diff
@@ -174,6 +177,13 @@ std::unique_ptr<L0CommandList> L0Device::create_command_list() const {
   return std::make_unique<L0CommandList>(res);
 }
 
+unsigned L0Device::maxGroupCount() const {
```
Comment (alexbaden) [inline omniscidb/L0Mgr/L0Mgr.cpp:180]: Can we use an explicitly sized type here?

File: omniscidb/QueryEngine/Compiler/genx.ll [hunk 2]
```diff
@@ -50,3 +49,29 @@ define i64 @agg_count(i64* %agg, i64 %val) {
...
     ret i64 %old
 }
+
+define i64 @agg_count_skip_val(i64* %agg, i64 noundef %val, i64 noundef %skip_val) {
+    %no_skip = icmp ne i64 %val, %skip_val
+    br i1 %no_skip, label %.noskip, label %.skip
+.noskip:
+;;  todo: %old = call i64 @agg_count(i64* %agg, i64 %val)
+    %ld = load i64, i64* %agg
+    %old = atomicrmw add i64* %agg, i64 1 monotonic
```
Comment (alexbaden) [inline omniscidb/QueryEngine/Compiler/genx.ll:59]: do we need this atomic?

File: omniscidb/QueryEngine/Descriptors/QueryMemoryDescriptor.cpp [hunk 3]
```diff
@@ -1059,8 +1059,13 @@ bool QueryMemoryDescriptor::blocksShareMemory() const {
 }
 
 bool QueryMemoryDescriptor::lazyInitGroups(const ExecutorDeviceType device_type) const {
+#ifdef HAVE_L0
+  return false; /* L0 path currently does not provide lazy group by buffers initialization
```
Comment (alexbaden) [inline omniscidb/QueryEngine/Descriptors/QueryMemoryDescriptor.cpp:1063]: Won't this disable lazy init groups for CUDA too? Or are we not planning to build CUDA + L0?

Comment (kurapov-peter) [reply omniscidb/QueryEngine/Descriptors/QueryMemoryDescriptor.cpp:1063]: There's no way of building CUDA + L0 currently, so CUDA is not affected.
>Or are we not planning to build CUDA + L0

We do, and there're some changes that minimize the gap we have now, but I have no immediate plans for enabling both in a single build right now. Over time it became much easier to enable, so we might want to do that soon.

File: omniscidb/QueryEngine/Compiler/genx.ll (hunk 2 above)
Comment (kurapov-peter) [reply omniscidb/QueryEngine/Compiler/genx.ll:59]: Not really, a regular `add` works just fine here. At least for non-grouped aggregations that do not use shared memory ;). I plan to revisit it once we have shared memory turned on (which I'm working on now).

File: omniscidb/L0Mgr/L0Mgr.cpp (hunk 1 above)
Comment (kurapov-peter) [reply omniscidb/L0Mgr/L0Mgr.cpp]: Done

**Review by alexbaden — APPROVED** [review]

**Pull request merged**